### ISSUE TITLE
Harddrive directory crash on load and start

### ISSUE THREAD

**Issue by henrikstengaard** [midwan/amiberry#752]
**Describe the bug**
When I start Amiberry, load a configuration and start it crashes and outputs attached "harddrive_directory_error.txt".
The crash is somewhat related to harddrive directories, which are /media/hstwb-self-install symlinks to /media/usbx directories.
Removing these after load makes Amiberry start, but with them it crashes.
I know this is context related to my setup, but I have attached my configuration "hstwb-installer-020-jit.uae", which could give helpful insight.

In summary I use following:
- Raspberry Pi 400 with Raspberry Pi OS.
- Amiberry 3.4 beta (2021-01-02) SDL2 + DispmanX compiled from dev branch. SDL2 v2.0.12 compiled from source.
- JIT, 68020+68882, 2mb chip, 16mb fast Z3.
- A1200 with Kickstart 3.1.
- DH0 Hardfile .hdf using PFS3 filesystem.
- DH1 Harddrive directory adding /home/pi/amiga/hdfs/hstwb/dh1 directory.
- DH2-DH4 harddrive directory adding /media/hstwb-self-install/x symlinks adding /media/usbx/x directories.

**To Reproduce**

1. Start Amiberry.
2. Load "hstwb-installer-020-jit.uae" configuration.
3. Click start.
4. See attached "harddrive_directory_error.txt".

**Expected behavior**

Load configuration "hstwb-installer-020-jit.uae" and start should not crash as it has been working fine with Amiberry v3.1.2.

**Setup:**

- OS: Raspbian GNU/Linux 10 (buster), Raspberry Pi OS Lite 2020-12-02.

- Hardware: Raspberry Pi 400 Rev v1.0

**Additional context**

In attempt to get more details about the error, I managed to find a workaround that makes it possible to start my configuration with all harddrive directories. This might be useful as well regarding fixing the issue.

Steps to workaround the issue:
1. Start Amiberry.
2. Load "hstwb-installer-020-jit.uae" configuration.
3. Switch to harddrives tab and remove harddrives DH2-DH4.
4. Click start.
5. Amiberry boots DH0 startup sequence, but without media harddrives DH2-DH4.
6. Press F12.
7. Click restart.
8. Press start.
9. Amiberry now boots DH0 startup sequence with media harddrives DH2-DH4.

[hstwb-installer-020-jit.uae.txt](https://github.com/midwan/amiberry/files/5782535/hstwb-installer-020-jit.uae.txt)
[harddrive_directory_error.txt](https://github.com/midwan/amiberry/files/5782540/harddrive_directory_error.txt)

**Comment by henrikstengaard** [comment]
I did some additional testing and have attached a zip file with .uae configuration files and a blank .hdf used for testing.

Here's what I did to reproduce the error:
1. Attach a Fat32 USB stick/drive and have it mounted with usbmount.
2. Start Amiberry.
3. Created A1200 with 68030, 2mb chip, 16mb fast Z3, JIT.
4. Add blank "dh0.hdf" hardfile as bootable.
5. Save configuration as "a1200-new-jit1.uae".
6. Click start and Amiberry starts fine in AmigaDOS at DH0.
7. Press F12 and quit.
8. Start Amiberry again.
9. Load configuration "a1200-new-jit1.uae".
10. Add harddrive directory to /media/usbx/any-directory.
11. Save configuration as "a1200-new-jit2.uae".
12. Click start and Amiberry crashed with following output:

```
INFO: --- New exception ---
INFO: Error not in JIT code.
INFO: Segmentation Fault
INFO: info.si_signo = 11
INFO: info.si_errno = 0
INFO: info.si_code = -6
INFO: info.si_addr = 00000717
INFO: r0  = 0x00000000
INFO: r1  = 0xac5bc89c
INFO: r2  = 0x00000000
INFO: r3  = 0x00000008
INFO: r4  = 0xb6ff7968
INFO: r5  = 0x0000000b
INFO: r6  = 0xac5bc89c
INFO: r7  = 0x000000af
INFO: r8  = 0xbeff71ca
INFO: r9  = 0xac5bd450
INFO: r10 = 0x00000000
INFO: FP  = 0xac5bcf8c
INFO: IP  = 0x00000000
INFO: SP  = 0xac5bc898
INFO: LR  = 0x00000000
INFO: PC  = 0xb6de58ec
INFO: CPSR = 0x00000010
INFO: Fault Address = 0x00000000
INFO: Trap no = 0x0000000e
INFO: Err Code = 0x80000207
INFO: Old Mask = 0x0e017007
INFO: LR - 0x00000000: symbol not found
INFO: Stack trace:
INFO: 0x0024faec <(null) + 0x0024faec> (./amiberry)
INFO: 0xb692f130 <__default_rt_sa_restorer + 0x00000000> (/lib/arm-linux-gnueabihf/libc.so.6)
INFO: Stack trace (non-dedicated):
INFO: ./amiberry() [0x24fc04]
INFO: /lib/arm-linux-gnueabihf/libc.so.6(__default_rt_sa_restorer+0) [0xb692f130]
INFO: End of stack trace.
INFO: --- end exception ---
```

[amiberry_harddrive_directory_error.zip](https://github.com/midwan/amiberry/files/5787747/amiberry_harddrive_directory_error.zip)

**Comment by henrikstengaard** [comment]
Today I tried compiling prevoius versions from master branch and ran my configuration to get an idea of which versions, this issue started to occur.

Running prevoius Amiberry versions gave following results:

- v3.2: Load and start configuration crashes Amiberry.
- v3.2-beta.2: Works fine, load and start configuration. No crash.
- v3.2-beta.1: Works fine, load and start configuration. No crash.
- v3.1.3.1: Works fine, load and start configuration. No crash.
- v3.1.3: Works fine, load and start configuration. No crash.
- v3.1.2: Works fine, load and start configuration. No crash.

Something regarding handling of harddrives added as directories has changed between version v3.2-beta.2 and v3.2 that makes Amiberry crash in situations like mine.

**Comment by midwan** [comment]
@henrikstengaard 
Thanks, I haven't had time to look into this yet, as I have my hands full with another area I'm working on.
But I'll look into it once I'm done there, and get back to you.

**Comment by henrikstengaard** [comment]
Sure, no problem. I hope my comments will help you pin point where to look when you get around to it.

I'm going to take a look my-self examining changes between the two versions where it starts crashing and see if I can makes something out of it.

**Comment by midwan** [comment]
@henrikstengaard 
I did a quick test today:
- Used my existing WB3.9 installation
- Added an external FAT32 formatted USB stick, as a secondary drive (Directory)

I didn't get any crashes or otherwise problems with this approach. I could access the files normally under Workbench etc.

Maybe this is an issue with your symlinks? Have you tried using the direct path instead, to verify if that's the case?

Also, the config file you attached above is coming from an older version of Amiberry. If you're testing with the latest one, I highly recommend you recreate that config using that version instead, as there are certain differences (and bugs in older version files), which in some cases cause issues in my experience.

**Comment by henrikstengaard** [comment]
I will see what I can do in order to re-create it. We just need to re-create from the same steps.

Did you use a PFS3 handler for your HDF file, a HDF file per partition?
Did you use HDF with RDB, HDF with multiple partitions?
Did you use FFS or PFS3 or other filesystem?
Or harddrives added as directories?

I find it a bit troublesome having to potentially re-create config files for each new version of Amiberry.

But I will retry with latest version and base it on steps that can be done with a clean setup.

**Comment by henrikstengaard** [comment]
I downloaded latest Amiberry BETA 2021-02-15 and managed to create a configuration from scratch that results in the same crash.

It has the following hard drive setup:
- DH0: Hardfile /home/pi/amiga/test.hdf (100mb HDF created using Create Hardfile under Hard drives/CD).
- DH2: Directory /media/usb/amigaos.
- DH1: Directory /home/pi.

I have attached my configuration and hardfile zipped. When I start Amiberry, load configuration "amiberry-beta-2021-02-15-crash.uae" and click start Amiberry crashes intiantly with similar error message as showed earlier.

**Observation 1**: If I start Amiberry, load configuration "amiberry-beta-2021-02-15-crash.uae", remove either hard drive DH1 or DH2 and click start then Amiberry starts fine.

**Observation 2**: If I start Amiberry, load configuration "amiberry-beta-2021-02-15-crash.uae", remove both hard drive DH1 and DH2 and click start then Amiberry starts fine. Now if I press F12 and load configuration "amiberry-beta-2021-02-15-crash.uae" again and click reset, then Amiberry starts fine with all configured hard drives DH0, DH1 and DH2. A Workbench adf is required to verify hard drives are present.

I simply can't get my head around why my attached configuration "amiberry-beta-2021-02-15-crash.uae" crashes as it doesn't contain anything special (blank hdf and simple directories) and configuration is created from scratch using latest dev Amiberry BETA 2021-02-15. Again this setup worked fine with earlier.

With the attached configuration and hdf, I hope you have some detail to re-create the crashing setup and/or investigate it.
[amiberry-beta-2021-02-15-crash.zip](https://github.com/midwan/amiberry/files/6029228/amiberry-beta-2021-02-15-crash.zip)

**Comment by henrikstengaard** [comment]
I also tried with a different FAT32 usb stick and it didn't have any effect and still results in a crash

**Comment by midwan** [comment]
@henrikstengaard 
Thanks for the repro steps and files.
I'll take another look at this over the next few days.

**Comment by Yserra** [comment]
When I had a similar issue, the problem was 68040.library, or any optional 680x0.library that activates the MMU (from MMULibs, from an accelerator card...). The JIT crashes when the MMU is activated. Just delete these libraries.

**Comment by henrikstengaard** [comment]
My hdf files are blank or doesn't contain any boot or system files and never gets to loading 680x0 libraries. My crash happens before Amiberry attempts to start using configured hard drives

**Comment by midwan** [comment]
@henrikstengaard 
I fixed some filesystem related bugs recently (after your tests above), which resulted in crashes if dealing with filenames which contained invalid characters (especially when trying to save to an NTFS formatted USB stick, for example). 

I'm not sure if that was related to your problem, but doing a quick test here with your hardfile and a similar set-up (2 extra directories mapped as drives, one being the home directory), did not show any crashes for me. At least, not yet. I'll keep trying with a few different combinations, to see if I can find some detail that I missed and might cause this.

But perhaps meanwhile you could also give it a try with the latest version, after the bug fixes I mentioned above? It would be great if that fixed this problem as well. :)

**Comment by henrikstengaard** [comment]
I tried latest BETA 2021-03-03 and it still crashes when i try to start any of my configurations that has hard drives added to directory /media/usbx.

However I made another observation today and I noticed that Amiberry only crashes the very first time I start Amiberry loaded with a configuration that has /media/usbx harddrives added. If I do the following steps first before loading and starting any configuration, all my configurations work fine:
1. Start Amiberry.
2. Switch to ROM tab.
3. Select AROS Ks ROM (built-in).
4. Click Start.
5. Press F12.
6. Click Restart (mouse cursor disappear every time I click Restart, I will create another issue for this).
7. Press ESC.
8. Press F12 (this restores mouse cursor).

After these steps any of my configurations work fine without crashes. It seems that Amiberry is in some sort of state that makes it crash before the actual Amiga emulation has startet.

**Comment by midwan** [comment]
@henrikstengaard 
Can you please try with the latest from `master` ? There was a bug fixed that might have affected this, so I'm curious to see if you can still recreate the problem.

**Comment by henrikstengaard** [comment]
Sure, i will give latest master commit/release a try

**Comment by midwan** [comment]
@henrikstengaard 
Any luck with this? Did you get a chance to try it again?

**Comment by henrikstengaard** [comment]
I updated to latest master commit today and compiled source code for KMS, OpenGL, X11. In about i see v4.2.0 (2021-12-16) and linked against sdl2 v2.0.14, but no matter what i configure i only see a black screen when clicking start button. I see Amiberry setup gui where can configure CPU, hard drives etc., but nothing when trying to start.

Has something changed display wise since v3.x i need to configure for v4.2.0? I'm still using the setup described in the issue

**Comment by midwan** [comment]
If you're using KMSDRM as your back-end, make sure the sdl2 thread option is _disabled_ in `amiberry.conf`. Otherwise you'll only get a black screen as you describe.

**Comment by henrikstengaard** [comment]
If i set option use_sdl2_render_thread=no in my conf, Amiberry crashes with a segmentation fault before showing any gui

**Comment by solskogen** [comment]
I'm quite sure your SDL installation is broken. Can you check if you have multiple versions of SDL installed? You might want to follow the guide on how to install the latest SDL on Raspberry Pi OS on the amiberry wiki.

**Comment by henrikstengaard** [comment]
@solskogen Sure, might be time to make a clean install of Raspbian and retry.

With a raspberry pi 4 / 400 and Raspbian, is KMSDRM the recommended solution to install/compile?

**Comment by solskogen** [comment]
Not really. Yes, you might gain /some/ performance, but not much. If you use the normal desktop with Raspberry Pi OS (I would recommend 64bit unless you have specific needs for 32bit) you don't need to install SDL manually either, just use the one that comes with Raspberry Pi OS. You only need to compile it yourself if you need KMSDRM.

**Comment by henrikstengaard** [comment]
I tried again this time with a clean download of Raspberry Pi OS Lite, but still resulting in segmentation fault.

Here's what I did:
- update raspberry pi os.
- install dev packages for compiling sources.
- cloned git repo.
- compiled sources: make -j2 PLATFORM=rpi4-sdl2

With use_sdl2_render_thread=no i get segmentation fault when loading my configuration and click start.

With use_sdl2_render_thread=no outcommented i get "double free or corruption (fasttop) aborted " when loading my configuration and click start.

Amiberry starts fine if i just run Amiberry and click start without any configuration.

I still have to remove any /media/x hard drives from my configuration before I can start without crash and segmentation fault.

My workaround as described perviously is still needed to get my configuration starting.

**Comment by solskogen** [comment]
With lite you need an updated SDL version.

ons. 29. des. 2021, 01:21 skrev Henrik Nørfjand Stengaard <
***@***.***>:

**Comment by henrikstengaard** [comment]
Is latest Amiberry release compiled against latest sdl2 v2.0.18

**Comment by henrikstengaard** [comment]
I downloaded and compiled SDL2 v2.0.18 using the steps you had described once in your wiki with arguments ```./configure --enable-video-kmsdrm --disable-video-rpi && make -j$(nproc) && sudo make install``` and compiled latest Amiberry master branch with arguments ```make -j2 PLATFORM=rpi4-sdl2``` and the issue is now gone without crash or segmentation fault 👍.

I'm surprised SDL2 has such impact on my issue.

I tested this using latest Raspberry Pi OS Lite and it seems something has changed regarding it's video driver. It's no longer possible to which between fake KMS and KMS.
